api/oc_push.c

```diff
@@ -0,0 +1,2605 @@
+
+	   /* oc_rep_close_object(parent, key) */
+	   g_err |= oc_rep_encoder_close_container(parent, &child);
+		break;
+
+	case OC_REP_OBJECT_ARRAY:
+
+		/* oc_rep_open_array(root, xxxx) */
+	   g_err |= oc_rep_encode_text_string(parent, oc_string(rep->name), oc_string_len(rep->name));
+	   memset(&child, 0, sizeof(child));
+	   g_err |= oc_rep_encoder_create_array(parent, &child, CborIndefiniteLength);
+
+	   /* recurse remaining objects... */
+	   obj = rep->value.object_array;
+	   while (obj)
+	   {
+			do
+			{
+				/* oc_rep_object_array_begin_item(key) */
+				CborEncoder obj_map;
+				memset(&obj_map, 0, sizeof(obj_map));
+				g_err |= oc_rep_encoder_create_map(&child, &obj_map, CborIndefiniteLength);
+
+				_build_rep_payload(&obj_map, obj->value.object);
+
+				/* oc_rep_object_array_end_item(key) */
+				g_err |= oc_rep_encoder_close_container(&child, &obj_map);
+			} while (0);
+			obj = obj->next;
+	   }
+
+	   /* oc_rep_close_array(root, xxxx); */
+	   g_err |= oc_rep_encoder_close_container(parent, &child);
+		break;
+
+	default:
+		break;
+	}
+
+	_build_rep_payload(parent, rep->next);
+
+	return;
+}
+
+
+
+static oc_pushd_rsc_rep_t * _find_pushd_rsc_rep_by_uri(oc_string_t *uri, size_t device_index)
+{
+	oc_pushd_rsc_rep_t *pushd_rsc_rep = (oc_pushd_rsc_rep_t *)(oc_list_head(pushd_rsc_rep_list));
+
+	while (pushd_rsc_rep)
+	{
+		if (!strcmp(oc_string(pushd_rsc_rep->resource->uri), oc_string(*uri))
+				&& (pushd_rsc_rep->resource->device == device_index))
+		{
+			break;
+		}
+		else
+		{
+			pushd_rsc_rep = pushd_rsc_rep->next;
+		}
+	}
+
+	return pushd_rsc_rep;
+}
+
+
+
+static void get_pushd_rsc(oc_request_t *request, oc_interface_mask_t iface_mask, void *user_data)
+{
+	(void)user_data;
+
+	int result = OC_STATUS_OK;
+	oc_pushd_rsc_rep_t *pushd_rsc_rep = _find_pushd_rsc_rep_by_uri(&request->resource->uri, request->resource->device);
+
+	if (!pushd_rsc_rep)
+	{
+		OC_PUSH_ERR("something wrong, can't find resource representation for pushed resource (%s)...",
+				oc_string(request->resource->uri));
+		return;
+	}
+
+	if (pushd_rsc_rep->rep)
+	{
+		oc_rep_begin_root_object();
+		switch (iface_mask)
+		{
+		case OC_IF_BASELINE:
+			oc_process_baseline_interface(request->resource);
+			/* fall through */
+		case OC_IF_R:
+		case OC_IF_RW:
+			_build_rep_payload(&root_map, pushd_rsc_rep->rep);
+			break;
+		default:
+			break;
+		}
+		oc_rep_end_root_object();
+
+		oc_send_response(request, result);
+	}
+	else
+	{
+		OC_PUSH_ERR("resource representation for pushed resource (%s) is found, but no resource representation for it is built yet!",
+				oc_string(request->resource->uri));
+
+		oc_send_response(request, OC_STATUS_NOT_FOUND);
+	}
+
+	return;
+}
+
+
+
+/*
+ * @brief				check if "rt" of pushed resource is part of "rts" (all value of "rt" should be part of "rts")
+ *
+ * @param recv_obj
+ * @param rep
+ * @return	not 0: found
+ * 			0: not found
+ */
+static char _check_pushd_rsc_rt(oc_recv_t *recv_obj, oc_rep_t *rep)
```

Comment:
(CheckPoint 1: checked this file upto this point, publishing my review so changes can be done, will read the rest tomorrow.)

Reply:
oops.. fixed...